src/util/evalUnits.ts

```diff
@@ -53,19 +81,26 @@ export enum EvalMessages {
  * the first is the actual calculated number (or null if none) while
  * the second is the message that should appear in case of error
  */
-export function evalUnits(input: string): [number | null, string] {
+export function evalUnits(input: string): [BN | number | null, string] {
```

Comment:
This function has become extremely complex and I don't think I fully get why. We should simply be always returning `BN`.

Take a look at my proposal:
```js
function evalUnits(input: string) -> [BN | null, string | null] {
   const symbol = input.replace(/[0-9,. ]/g, '');
   const siVal = si.find(s => s.symbol === symbol);
   if (symbol && !siVal) {
      return [ null, 'Invalid symbol' ];
   }

   const numberStr = input.replace(symbol, '').replace(',', '.);
   const asFloat = parseFloat(numberStr);
   if (asFloat.toString() === numberStr) {
      // now that's the tricky part - how do we convert a float into a BN?
      // let's move the comma as far as we can to the right :)
      const [decPart, fracPart] = numberStr.split('.');
      const decimals = fracPart.length;
      const exp = new BN(10).pow(new BN(decimals));
      const number = new BN(decPart).mul(exp).add(new BN(fracPart));
      const withSi = number.mul(siVal.value).div(exp);
      return [ withSi, null ];
   }

   // we can't parse the float correctly - perhaps it's above the precision. Let's simply try BN.
   try { 
      const bn = new BN(numberStr);
      return [ bn.mul(siVal.value), null];
   } catch (e) {
      return [null, `Invalid number: ${e}`]
   }
}
```

Note that we should keep `siVal.value` as `BN` already. Best to do `new BN(10).pow(new BN(21))` (instead of `1e21`) because parsing large floats fails as you've noticed.

Reply:
Was trying to make the function more details with specific errors and not generic (e.g. you dont separate Negative errors with e.g. zero) - but I get the point that it became very complicated for no actual reason.

Reply:
This COULD be fixed if the final outcome would ALWAYS be multiplied with the chain.decimals (which I will give it a try)